compiler/qsc_eval/src/lib.rs

```diff
@@ -1448,6 +1462,31 @@ impl State {
             span,
         }
     }
+
+    fn counting_call(&mut self, name: &str, arg: Value) -> Value {
+        let callable = if let Value::Closure(closure) = arg {
+            (closure.id, closure.functor)
+        } else {
+            arg.unwrap_global()
+        };
+        match name {
```

Comment:
If we are just using the callable name to determine whether it's a counting call, it is possible that a user creates a callable with the same name, the callable would not behave as the user expects it, right?

It is arguably a very rare case, but maybe it could be even more rare if we also checked more properties like making sure `CallableImpl` is the `Intrinsic` variant, `CallableKind` is the `Operation` variant and possibly also the `input` and `output` properties. This would make it extremely unlikely that someone would implement operations with the exact name and properties as these ones.

Reply:
I just realized that this comment would be better placed at the `is_counting_call` function!

Reply:
This is an interesting question. The way the match is used up on line 979 already only checks the name of the callable if it is intrinsic, so that part is covered. At first I figured that would be sufficient because the compiler enforces that all intrinsic callables are globally unique (see `Qsc.Resolve.DuplicateIntrinsic` error in the resolver). But this does change a bit when it comes to dependencies on external projects, as a transitive dependency's globals are not included in the global table for a compilation such that duplicate intrinsic functions can exist. To put it another way, if A depends only on B, and B depends only on C, A can declare an intrinsic with the same name as one declared in C without it triggering this error because global items in C are not visible in A. This gets further complicated by simulatable intrinsics which are not covered by this check. I'll file a follow up issue to cover closing that gap and checking whether there is a reasonable way to ensure globally unique intrinsics even across transitive dependencies.

The good news is that in practice this won't cause a problem for any of these callables, as they are in the standard library and are currently included in every compilation, so anyone declaring an intrinsic callable with the same will get an error. But if/when we make skipping the stdlib an option we have to consider how that will interact with some of these other intrinsics. Like some of the type conversion intrinsics, I think this might make sense to include in the non-quantum part of the stdlib (core or otherwise) that cannot be disabled. For now, the check is safe as there can only be one callable that is both intrinsic and has these names.

Reply:
Filed https://github.com/microsoft/qsharp/issues/1817